DRA 1.32: beta (#48330)
kubernetes/website: content/en/docs/concepts/scheduling-eviction/dynamic-resource-allocation.md

Reviewer: Try adding this into the [front matter](https://gohugo.io/content-management/front-matter/):
```yaml
api_metadata:
- apiVersion: "resource.k8s.io/v1beta1"
  kind: "ResourceClaim"
- apiVersion: "resource.k8s.io/v1beta1"
  kind: "ResourceSlice"
```
etc

You might need to manually generate placeholder beta API docs to make that work. If you do, mention in this PR that you have done that, and _don't_ mark the placeholder pages as autogenerated (again, see front matter).

To see how this renders, look at the top of the right hand nav for https://kubernetes.io/docs/concepts/storage/persistent-volumes/

Author: Added, seems to work: https://deploy-preview-48330--kubernetes-io-main-staging.netlify.app/docs/concepts/scheduling-eviction/dynamic-resource-allocation/

I noticed that `DeviceClass` is under `extended-resources`. I had to leave it there also for the v1beta1 version because moving it manually left a dangling link. This needs to be fixed when preparing the next manual update.

cc @tengqm

You may also be able to mutate the incoming Pod, at admission time, to unset
the `.spec.nodeName` field and to use a node selector instead.

> ## Admin access

Reviewer: How about:
```suggestion
## Resources with privilege escalation risks {#admin-access}
```

Author: I prefer the shorter "Admin access" because "Resources with privilege escalation risks" focuses too much on the risk (IMHO) and isn't quite correct (it's about requests, not the resources).

{{< feature-state feature_gate_name="DRAAdminAccess" >}}

> 

Reviewer: ```suggestion

You can mark a ResourceClaimTemplate as providing privileged features.

```

Author: Added, with one small tweak: "... mark` a *request* in a ResourceClaim *or* ResourceClaimTemplate ... as *having* ...".

It's per-request, not per claim (I wasn't precise there myself earlier, fixed), and it's not just at runtime ("providing") that such a request is special. It's also during scheduling.

A ResourceClaim with admin access grants access to devices which are in use and
may enable additional permissions when making the device available in a
container:

```yaml
apiVersion: resource.k8s.io/v1beta1
kind: ResourceClaimTemplate
metadata:
  name: large-black-cat-claim-template
spec:
  spec:
    devices:
      requests:
      - name: req-0
        deviceClassName: resource.example.com

>         adminAccess: true

Reviewer: This would be easier to document if the field were named, eg, `considerAsPrivileged` or `privileged` or `limitedAccess`.

Author: cc @thockin @klueska 

For consideration in future APIs...

```

If this feature is disabled, the `adminAccess` field will be removed
automatically when creating such a ResourceClaim.

> Admin access is a privileged mode which should not be made available to normal
> users in a multi-tenant cluster. Cluster administrators can restrict usage of
> this features by installing a validating admission policy similar to
> [the in-tree example](https://github.com/kubernetes/kubernetes/blob/33ea278/test/e2e/dra/test-driver/deploy/example/admin-access-policy.yaml)
> when enabling this feature.

Reviewer: ```suggestion
Setting `adminAccess` true marks the allocated resources as privileged: facilities
or devices that should not be made available to normal
users in a multi-tenant cluster. As a cluster administrator, you can restrict usage of
privileged resources by installing a ValidatingAdmissionPolicy similar to
the [in-tree example](https://github.com/kubernetes/kubernetes/blob/33ea278/test/e2e/dra/test-driver/deploy/example/admin-access-policy.yaml)
when enabling this feature.
```

However, at beta we expect the example ValidatingAdmissionPolicy to be copied into the docs repo (possibly with a light tweak), rather than linking to code from k/k.

Author: > Setting `adminAccess` true marks the allocated resources as privileged: facilities or devices that should not be made available to normal users in a multi-tenant cluster.

This is not accurate. Already the request is special (scheduling) and then the access to them in containers *for this request*. Elsewhere, the devices are normal and can be used by normal users.

Author: > However, at beta we expect the example ValidatingAdmissionPolicy to be copied into the docs repo (possibly with a light tweak), rather than linking to code from k/k.

Inlined as a YAML code block, with tweaked documentation for admins.

Author: https://deploy-preview-48330--kubernetes-io-main-staging.netlify.app/docs/concepts/scheduling-eviction/dynamic-resource-allocation/#admin-access

In addition to enabling the feature in the cluster, a resource driver also has to
be installed. Please refer to the driver's documentation for details.

### Enabling admin access

> Admin access is an *alpha feature* and only enabled when the

Reviewer: ```suggestion
[Admin access](#admin-access) is an *alpha feature* and only enabled when the
```

Author: Done.
